### ISSUE TITLE
OOM due to json.dumps, use alternative json parser or dump to file directly?

### ISSUE THREAD

**Issue by laundmo** [gaogaotiantian/viztracer#115]
OS: windows 10
Python: Python 3.9.5 64bit
RAM: 16GB
Version: VizTracer 0.13.2

Im trying to run viztracer on a relatively small project, and even this generates a OOM when dumping to html or json

Possible solutions:
 - dont use `json.dumps`, as this uses extra memory. use `json.dump` to stream to a file directly
 - Use a more memory efficient json parser
<details>
  <summary>Click to see error message</summary>

```
==================================================
== Starting from version 0.13.0, VizTracer will ==
== use json as the default report file. You can ==
== generate HTML report with "-o result.html"   ==
Dumping trace data, total entries: 2272769
Traceback (most recent call last):
  File "d:\prog\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "d:\prog\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "D:\prog\Python37\Scripts\viztracer.exe\__main__.py", line 7, in <module>
  File "d:\prog\python37\lib\site-packages\viztracer\main.py", line 391, in main
    success, err_msg = ui.run()
  File "d:\prog\python37\lib\site-packages\viztracer\main.py", line 213, in run
    return self.run_command()
  File "d:\prog\python37\lib\site-packages\viztracer\main.py", line 293, in run_command
    return self.run_code(code, main_mod.__dict__)
  File "d:\prog\python37\lib\site-packages\viztracer\main.py", line 246, in run_code
    self.exit_routine()
  File "d:\prog\python37\lib\site-packages\viztracer\main.py", line 377, in exit_routine
    self.save(self.tracer)
  File "d:\prog\python37\lib\site-packages\viztracer\main.py", line 371, in save
    tracer.save(output_file=ofile, save_flamegraph=options.save_flamegraph, file_info=True)
  File "d:\prog\python37\lib\site-packages\viztracer\viztracer.py", line 172, in save
    rb.save(output_file=output_file, file_info=file_info)
  File "d:\prog\python37\lib\site-packages\viztracer\report_builder.py", line 135, in save
    data = self.generate_json(allow_binary=True, file_info=file_info)
  File "d:\prog\python37\lib\site-packages\viztracer\report_builder.py", line 108, in generate_json
    return json.dumps(self.combined_json)
  File "d:\prog\python37\lib\json\__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "d:\prog\python37\lib\json\encoder.py", line 202, in encode
    return ''.join(chunks)
MemoryError
```

</details>

**Comment by gaogaotiantian** [comment]
This is something that we could improve(potentially a lot). ``dumps`` is not a good solution memory wise, and we should change that. It was there because the default output format used to be ``html`` and that requires json string. There are some optimizations in internal data storage as well. I'll look into it.

I also noticed that you have 2 mil entries. Did you changed the ``tracer_entries`` or this is a multi process collection?

**Comment by laundmo** [comment]
Sorry, yes, i did change tracer_entries for that particular run but the same OOM was happening on the default

**Comment by gaogaotiantian** [comment]
We should definitely support as many entries as possible by optimizing memory optimizations. I'm not accusing you of using more entries, that's the reason the option is there. I just want to make sure you did use the option.

This is a valid issue and I will make optimizations. The estimation of JSON file becomes inaccurate as we are adding stuff into the file. Also that's the estimation, even when it's written, was of the file size(which is kind of ambiguous in the docs and I should fix that), not RAM usage.

In any case, thank you for noticing this issue and I'll try to improve it asap.

**Comment by gaogaotiantian** [comment]
Something is off to me here. I know there's some deficiency in memory usage, but I've done some experiment myself and the result is kind of far off. Do you know how much memory did VizTracer use when it broke down? I know you have 16G RAM, but it could've been used by other programs(including the python program that you are using VizTracer to trace). In my experience, the duplicate JSON data should take about 1-2MB per 10k entries, which is about the same as the file size. So, even with 2 mil entries, it should take an extra 200-400MB RAM with ``dumps``.

Also, you said even with default entry number, you will get the MemoryError, but this is the first issue people have with insufficient memory. It's hard to imagine that if default entry number would easily cause 16G RAM to drain, people could live with it for such a long time.

Could you try to reproduce this without the other possible factors? For example, you could trace this:
```
def f():
    return

for _ in range(1000000):
    f()
```
which should use all the default circular buffer. Will this use up your RAM as well?

**Comment by laundmo** [comment]
@gaogaotiantian i was testing this on my work PC which is ancient, no issues on my home pc which is VERY good in comparison.

I will try the example tomorrow, when im back at work.

I was watching the memory usage in ProcessExplorer while it was running, and did not notice much of a spike, although that could be due to the fact the memory usage was only updated every ~1s

**Comment by gaogaotiantian** [comment]
So I was digging into this issue and found/fixed something.
- I optimized the name string storage, which should result in ~ 20% memory saving
- I realized that ``json.dump`` is 3x slower than ``json.dumps`` because it's using pure python instead of c
- I won't put any third party library in VizTracer for now, but ``orjson`` is always supported as an optional json parser. However, ``orjson`` does not support ``dump``.
- As of now, the default behavior will stay the same(except for the extra memory saving from name string in C code), which is dumping to a string then store to file.
- However, I added an option ``--minimize_memory`` to use ``json.dump`` to dump json data iteratively, which should potentially save some memory when it matters(and of course cost more time)
Is this something you think sufficient? Or do you have any other requests?

The changes is checked in and should be merged to master soon.

**Comment by gaogaotiantian** [comment]
The changes are merged in in #116. This feature will be there for next release. If you want to try it now, you can build from source by ``pip install git+https://github.com/gaogaotiantian/viztracer``
